tmt/steps/provision/__init__.py

```diff
@@ -479,6 +408,229 @@ def _export_environment(environment):
+        Reboot the guest, return True if successful
+
+        Parameter 'hard' set to True means that guest should be
+        rebooted by way which is not clean in sense that data can be
+        lost. When set to False reboot should be done gracefully.
+        """
+
+        raise NotImplementedError()
+
+    def reconnect(self):
+        """ Ensure the connection to the guest is working after reboot """
+        # Try to wait for machine to really shutdown sshd
+        # TODO: rename the constant - reconnect() in this form works well even
+        # for ssh-less guests, so there's no point in calling it SSH_*.
+        time.sleep(SSH_INITIAL_WAIT_TIME)
+        self.debug("Wait for a connection to the guest.")
+        for attempt in range(1, CONNECTION_TIMEOUT):
+            try:
+                self.execute('whoami')
+                break
+            except tmt.utils.RunError:
+                self.debug('Failed to connect to the guest, retrying.')
+                time.sleep(1)
+
+        if attempt == CONNECTION_TIMEOUT:
+            self.debug("Connection to guest failed after reboot.")
+            return False
+        return True
+
+    def remove(self):
+        """
+        Remove the guest
+
+        Completely remove all guest instance data so that it does not
+        consume any disk resources.
+        """
+        self.debug(f"Doing nothing to remove guest '{self.guest}'.")
+
+    def _check_rsync(self):
+        """
+        Make sure that rsync is installed on the guest
+
+        On read-only distros install it under the '/root/pkg' directory.
+        Returns 'already installed' when rsync is already present.
+        """
+
+        # Check for rsync (nothing to do if already installed)
+        self.debug("Ensure that rsync is installed on the guest.")
+        try:
+            self.execute("rsync --version")
+            return "already installed"
+        except tmt.utils.RunError:
+            pass
+
+        # Check the package manager
+        self.debug("Check the package manager.")
+        try:
+            self.execute("dnf --version")
+            package_manager = "dnf"
+        except tmt.utils.RunError:
+            package_manager = "yum"
+
+        # Install under '/root/pkg' for read-only distros
+        # (for now the check is based on 'rpm-ostree' presence)
+        # FIXME: Find a better way how to detect read-only distros
+        self.debug("Check for a read-only distro.")
+        try:
+            self.execute("rpm-ostree --version")
+            readonly = (
+                " --installroot=/root/pkg --releasever / "
+                "&& ln -sf /root/pkg/bin/rsync /usr/local/bin/rsync")
+        except tmt.utils.RunError:
+            readonly = ""
+
+        # Install the rsync
+        self.execute(f"{package_manager} install -y rsync" + readonly)
+
+    @classmethod
+    def requires(cls):
+        """ No extra requires needed """
+        return []
+
+
+class GuestSSH(Guest):
```

Comment:
Naming nitpick: Should we go with `GuestSsh` instead of `GuestSSH` to keep an always-camel-case consistency with for example `DiscoverFmf` and `ReportHtml`? Or is this a better approach for abbreviations?

Reply:
Force of habit, I guess. In my world, `SSH` is an abbreviation and therefore not a subject to camel-case (`GuestSSH`, GuestSSHWithSomethingExtra`, etc.). But since there are already `ReportHtml`, it makes sense to remain consistent, `GuestSsh` it shall be.

Reply:
Fixed in 59016c27a30536f5b83a2fad345cfe62ecf47ce6.